Introduce SQL functions for computing hash sums

mxm opened the pull request:
This addresses issue #5403.

crate-jenkins commented:
Can one of the admins verify this patch?

seut commented on sql/src/test/java/io/crate/operation/scalar/HashFunctionsTest.java:
```diff
@@ -0,0 +1,81 @@
+
+    @Test
+    public void testChecksums() throws Exception {
+        assertEvaluate("md5('©rate')", "53cee571b9fbab07cc894d55988cc70b");
```
assertNormalize should be used when passing in literals

mxm replied:
Assuming that `assertEvaluate` does a normalize (it does), this should be fine. Still better to test normalize explicitly. I added a test case.

seut commented on blackbox/docs/sql/scalar.txt:
```diff
@@ -215,6 +215,41 @@ locale-sensitive or context-sensitive mappings.
+-----------------------------
+
+Computes the SHA2-256 checksum of the given string.
+See sha1 for an example.
```
pls add a RST link to sha1

mxm replied:
Thanks, will add the link.

matriv commented on blackbox/docs/sql/scalar.txt:
```diff
@@ -215,6 +215,41 @@ locale-sensitive or context-sensitive mappings.
+::
+
+    cr> select sha1('crate')
+    ... limit 1;
```
limit here is unnecessary or?

mxm replied:
Yep! Thanks :)

dobe commented on sql/src/main/java/io/crate/operation/scalar/string/HashFunctions.java:
```diff
@@ -0,0 +1,102 @@
+        public BytesRef digest(BytesRef input) {
+            final MessageDigest messageDigest;
+            try {
+                messageDigest = MessageDigest.getInstance(method);
```
consider using org.elasticsearch.common.hash.MessageDigests to prevent creating a new instance for every row

mxm replied:
Good point. I thought it would return the same instance but it creates a new one for each call. Very expensive.

mxm replied:
`MessageDigests` doesn't have a function for SHA512. I've removed SHA512. We could add the function to `MessageDigests` in an upstream patch.

dobe commented on blackbox/docs/sql/scalar.txt:
```diff
@@ -215,6 +215,41 @@ locale-sensitive or context-sensitive mappings.
+    +------------------------------------------+
+    SELECT 1 row in set (... sec)
+
+sha256('string') returns string
```
i am not sure if we should introduce another new function for 512 and 256, mysql has sha2(string, l) which i would prefer here

mxm replied:
I find it more convenient to have another function rather than parameterizing. If we want to comply with MySQL, then we should change the function.

mxm replied:
I'm opting against the `SHA2(x, len)` because we currently only support SHA256. I removed SHA512 because I didn't want to duplicate the upstream `MessageDigests` code. We can add a `SHA2(x, len)` function once we support different SHA2 sizes.

dobe replied:
ok, so i propose we then remove also sha256 in order to get this in. there are other crypto functions as well in future and we have not decided on an interface for them, so i would stick with sha1 and md5 since those are known functions at least from mysql and postgresql (md5)

reagding the upstream code ... internal APIs should not lead our decision on user interface.

mxm replied:
You're right, internals APIs should not dictate what is possible for the user. So let's revisit the SHA2 hash functions interface discussion at a later point in time and keep only SHA1 and MD5 for now. Perhaps users will provide some feedback on how the interface should look like when they start using the new hash functions.

matriv commented:
@mxm Please also add an entry in CHANGES.txt: https://github.com/crate/crate/blob/master/CHANGES.txt#L20 for the new feature.

mxm commented:
Thanks @matriv. Added a note to the change log.

mxm commented:
Thank you for the reviews @seut @dobe @matriv. I think I addressed all of your comments. Let me know if I can add anything. I will squash the commits once this is good to go.

matriv commented on CHANGES.txt:
```diff
@@ -18,6 +18,9 @@ Breaking Changes
 Changes
 =======
 
+ - Added support for hash sums (MD5, SHA1, SHA256). Please
```
Please change to: **Implemented hash sum scalar functions (MD5, SHA1, SHA256).**

mxm replied:
Sure. Will change.

matriv commented on CHANGES.txt:
```diff
@@ -18,6 +18,9 @@ Breaking Changes
 =======
 
+ - Added support for hash sums (MD5, SHA1, SHA256). Please
+    see :ref:`sha1 <sha1>`.
```
This reference here won't work as CHANGES.txt is not part of the generated html documentation.

mxm replied:
I assumed this would work because I found a reference in the current CHANGES.txt: https://github.com/crate/crate/blob/master/CHANGES.txt#L18 Will remove the reference.

matriv replied:
@mxm I'm really sorry but you can revert this since the CHANGES.txt becomes part of the release notes (inside generated html docs).

mxm replied:
No problem :)

mxm commented:
Thanks for merging! :)
